### PR TITLE
Add Builder in favour of constructors for frame msgs

### PR DESCRIPTION
**Pull request by rfeistenauer**
## Description

These changes improve extendability and usability of the `monitoring_frame::Message` class.

- [x] Add Builder and remove constructors
- [x] Make all additional fields optional
- [x] Remove dubious `operator==`
- [x] Check & update tests

---

<details>
<summary>PR Checklist</summary>

### Things to add, update or check by the maintainers before this PR can be merged.

- [ ] Public api function documentation
- [ ] Architecture documentation reflects actual code structure
- [ ] [Tutorials](https://wiki.ros.org/pilz_robots/Tutorials/)
- [ ] Overview on [ROS wiki](https://wiki.ros.org/pilz_robots)
- [ ] Package Readme ([example pilz_robots](https://github.com/PilzDE/pilz_robots/blob/melodic-devel/README.md))
- [ ] Good commit messages ([some tips](https://dev.to/jacobherrington/how-to-write-useful-commit-messages-my-commit-message-template-20n9))
- [ ] CHANGELOG.rst updated
- [ ] Copyright headers
- [ ] Examples

### Review Checklist
- [ ] Soft- and hardware architecture (diagrams and description)
- [ ] Test review (test plan and individual test cases)
- [ ] Documentation describes purpose of file(s) and responsibilities
- [ ] Code (coding rules, style guide)

### Release planning (please answer)
- [ ] When is the new feature released?
- [ ] Which dependent packages do have to be released simultaneously?

### Hardware tests
_Unstrike the text below to enable automatic hardware tests if available. Otherwise use this as a request for the reviewer._
- [ ] ~Perform hardware tests~

</details>

### CODE REVIEW

File: standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp [hunk 1]
```diff
@@ -27,86 +28,95 @@
 #include "psen_scan_v2_standalone/data_conversion_layer/raw_scanner_data.h"
 
 #include "psen_scan_v2_standalone/data_conversion_layer/raw_data_array_conversion.h"
+#include "psen_scan_v2_standalone/util/matchers_and_actions.h"
+
+#define CHANGE_MSG_PROPERTY(msg, property, offset)                                                                     \
```
Comment (rfeistenauer) [inline standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp:33]: it seems to be more an increase, than an change msg property

File: standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp [hunk 2]
```diff
@@ -27,86 +28,95 @@
...
+            .scannerId(msg.scannerId())                                                                                \
+            .fromTheta(msg.fromTheta())                                                                                \
+            .resolution(msg.resolution())                                                                              \
+            .scanCounter(msg.scanCounter())                                                                            \
+            .activeZoneset(msg.activeZoneset())                                                                        \
+            .measurements(msg.measurements())                                                                          \
+            .intensities(msg.intensities())                                                                            \
+            .property(msg.property() + offset)
+
+#define CHANGE_MSG_VECTOR_PROPERTY(msg, property, index, offset)                                                       \
```
Comment (rfeistenauer) [inline standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp:44]: Same as above

File: standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp [hunk 3]
```diff
@@ -201,37 +204,49 @@ TEST(LaserScanConversionsTest, shouldThrowProtocolErrorOnMissingFrames)
...
+        measurements_it,
+        UnsafePointwiseDoubleItEq(stamped_msg.msg_.measurements().begin(), stamped_msg.msg_.measurements().end()));
+    EXPECT_THAT(
+        intensities_it,
+        UnsafePointwiseDoubleItEq(stamped_msg.msg_.intensities().begin(), stamped_msg.msg_.intensities().end()));
+    measurements_it += stamped_msg.msg_.measurements().size();
+    intensities_it += stamped_msg.msg_.intensities().size();
+  }
+  EXPECT_EQ(measurements_it, scan_ptr->getMeasurements().end());
+  EXPECT_EQ(intensities_it, scan_ptr->getIntensities().end());
```
Comment (rfeistenauer) [inline standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp:244]: What this test does is something like:

auto merged_measurements = MERGE_PROP_LISTS(measurements, stamped_msgs)
EXPECT_THAT(scan_ptr->measurements(), PointwiseDoubleEq(merged_measurements))

and the same for intensities? In tha case I would make 2 tests that do that in like 2 lines.

File: standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp [hunk 4]
```diff
@@ -240,57 +255,38 @@ TEST(LaserScanConversionsTest, laserScanShouldContainCorrectTimestampWhenBuildin
...
-  // Change last measurement value.
-  auto new_measurements2 = stamped_msgs[2].msg_.measurements();
-  new_measurements2[new_measurements2.size() - 1] += 10;
-  auto new_last_msg = copyStampedMsgWithNewMeasurements(stamped_msgs[2], new_measurements2);
-
-  // Build LaserScan with wrong message order.
+  auto stamped_msgs = createValidStampedMsgs(6);
+  CHANGE_MSG_VECTOR_PROPERTY(stamped_msgs.front().msg_, measurements, 0, 10.0);
+  CHANGE_MSG_VECTOR_PROPERTY(
+      stamped_msgs.back().msg_, measurements, stamped_msgs.back().msg_.measurements().size() - 1, 10.0);
```
Comment (rfeistenauer) [inline standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp:261]: maybe a define LAST_MSG_INDEX or something like that?

File: standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp [hunk 5]
```diff
@@ -240,57 +255,38 @@ TEST(LaserScanConversionsTest, laserScanShouldContainCorrectTimestampWhenBuildin
...
-  ASSERT_NO_THROW(scan_ptr.reset(new LaserScan{
-      data_conversion_layer::LaserScanConverter::toLaserScan({ stamped_msgs[1], new_last_msg, new_first_msg }) }););
+  ASSERT_NO_THROW(
+      scan_ptr.reset(new LaserScan{ data_conversion_layer::LaserScanConverter::toLaserScan(stamped_msgs) }););
 
-  // Assert order was corrected.
-  ASSERT_EQ(scan_ptr->getMeasurements().front(), new_measurements1.front());
-  ASSERT_EQ(scan_ptr->getMeasurements().back(), new_measurements2.back());
+  ASSERT_EQ(first_measurement, scan_ptr->getMeasurements().front());
+  ASSERT_EQ(last_measurement, scan_ptr->getMeasurements().back());
```
Comment (rfeistenauer) [inline standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp:271]: To clarify, that these are the relevant checks you could make them EXPECTs

File: standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp [hunk 6]
```diff
@@ -307,13 +303,12 @@ TEST(LaserScanConversionTest, conversionShouldIgnoreEmptyFramesForMonitoringFram
...
-    Stamped(MessageBuilder().fromTheta(Tenth(2000)).resolution(Tenth(2)).scanCounter(42).measurements({}), 8)
+    Stamped(createDefaultMsgBuilder().fromTheta(Tenth(2500)).measurements({}), 3),
+    Stamped(createDefaultMsgBuilder().fromTheta(Tenth(0)).measurements({}), 4),
+    Stamped(createDefaultMsgBuilder().fromTheta(Tenth(500)).measurements({}), 5),
+    Stamped(createDefaultMsgBuilder().fromTheta(Tenth(1318)).measurements({ 1., 2., 3. }), 6),
+    Stamped(createDefaultMsgBuilder().fromTheta(Tenth(1500)).measurements({}), 7),
+    Stamped(createDefaultMsgBuilder().fromTheta(Tenth(2000)).measurements({}), 8)
   };
 
   ASSERT_NO_THROW(data_conversion_layer::LaserScanConverter::toLaserScan(stamped_msgs));
```
Comment (rfeistenauer) [inline standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp:314]: I knwo you didn't change this but what exactly does this test ... test?
"conversionShouldIgnoreEmptyFramesForMonitoringFramesValidation"
... some frames with empty monitoring frames and different theta
"ASSERT_NO_THROW(data_conversion_layer::LaserScanConverter::toLaserScan(stamped_msgs));"

what would be different if they where not empty?

File: standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp (hunk 3 above)
Comment (martiniil) [reply standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp:244]: I misunderstood this test, it checked that all scan information is contained by checking the vector sizes. I think this is fine.

File: standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp (hunk 6 above)
Comment (martiniil) [reply standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp:314]: `conversionShouldIgnoreEmptyFramesWhenCheckingIfFromThetasFitTogether` - better?

**Pull request merged**